File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

void dfs(vector<int> &num, int target, int &ret, int ni, int now){
    // cout << ni << ":" << num[ni]  << " " << now << " " << s << "\n";
    if (ni == num.size()){
        if (now == target){
            // cout <<"RET\n";
            ret++;    
        }
        return;
    }
    dfs(num,target,ret,ni+1,now+num[ni]);
    dfs(num,target,ret,ni+1,now-num[ni]);
    
}

int solution(vector<int> numbers, int target) {
    //2^20
    int ret = 0;
    dfs(numbers, target, ret, 0, 0);
    return ret;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int dfs(int dep,int now,int[] numbers,int target){
        if (dep == numbers.length){
            if (now == target)
                return 1;
            else
                return 0;
        }else{
            return dfs(dep+1,now+numbers[dep],numbers,target)+dfs(dep+1,now-numbers[dep],numbers,target);
        }
    }
    
    public int solution(int[] numbers, int target) {
        int answer = dfs(0,0,numbers,target);
        return answer;
    }
}